Ambient policy enforcement (#15072)
istio/istio.io: content/en/docs/ambient/usage/policy/index.md

---

> title: Policy enforcement in ambient mode

Reviewer: Is this only Authz policy?

Reviewer: I would like to see this under a page for authz policy vs under usage/policy as if folks don't need authz policy, they don't need to care about this. Or if folks don't need L4 authz policy, they don't need to worry about this.

Reviewer 2: My suggestion is:

- this becomes a "Designing authorization policy for ambient mode" guide (or whatever other types of policy are different between L4 and L7 and need consideration)
- it is weighted to come after the initial L4/L7 feature introductions
- most of the content is here, and the L4/L7 feature introductions link out to this doc.

Author: ok, I thought we were attempting to get rid of a dedicated auth pol page for ambient. I'm ok with having it though.

Reviewer 2: We got rid of a dedicated page to move policy into the journey of an adopter:

- here's how you install
- here's how you can do L4 things
- here's how you can add on L7 things

We didn't want to leave policy to after you've learned about L7.
However, a page on how you should think about and design policy is a good thing to have, and by the time you get to reading that, you should have understood L4/L7/etc architecturally.

description: The two enforcement points for policy in an ambient mesh.
weight: 20
owner: istio/wg-networking-maintainers
test: no
---

> Istio's ambient data plane splits the data plane into two different sets of components. This architecture allows users to only pay for application layer processing if they need it. The trade-off made is the extra complexity of understanding how traffic flows through this architecture and where policy will be enforced. This guide will introduce a number of broad scenarios and explore the permutations therein and the details to understand for each scenario.

Reviewer: `extra complexity` - we need to spin positively. flexibility, pay only what you need

Author: I think we just delete this page. Right now it is basically just storing notes.

Reviewer: Ok, maybe good to have dedicated page for authz policy.

## TCP Enforcement

In the simplest enforcement scenario, you simply want to enforce policy against TCP attributes and you have no waypoint proxies in your traffic's path. These policies can be enforced by the ztunnel proxies.

Once you introduce a waypoint proxy, the ideal place to enforce policy shifts. Traffic arriving at the destination ztunnel will be coming from the waypoint's identity because waypoint proxies do not impersonate `src` identity on behalf of the client. This means that even if you only wish to enforce policy against TCP attributes, you should bind that policy to your waypoint proxy. A further TCP policy may be applied to your workload to request that ztunnel enforce things like, "in-mesh traffic needs to come from my waypoint in order to reach my application". This type of policy allows you to choose if "bypassing" the waypoint proxy is permissible in your scenario.

This table is based on the following invariants:

1. The source pod is a normal pod which has ztunnel enabled.
1. Redirection to the waypoint is configured correctly.
1. The waypoint is configured with the `istio.io/waypoint-for` label set to `service`.

> | Waypoint* | Attachment Style | Scope | Source Identity | Enforced By |
> | --- | --- | --- | --- | --- |
> | no | Selector | Pod | client pod | destination ztunnel |
> | yes | Selector | Pod | waypoint | destination ztunnel |
> | no | | Namespace | client pod | destination ztunnel |
> | yes | | Namespace | waypoint | destination ztunnel |
> | yes | `targetRefs` | Service | client pod | waypoint |
> | yes | `targetRefs` | Gateway | client pod | waypoint |
> 
> * Whether or not there is already a waypoint is in the traffic path.

Reviewer 2: ```suggestion
| Waypoint † | Attachment Style | Scope | Source Identity | Enforced By |
| --- | --- | --- | --- | --- |
| no | Selector | Pod | client pod | destination ztunnel |
| yes | Selector | Pod | waypoint | destination ztunnel |
| no | | Namespace | client pod | destination ztunnel |
| yes | | Namespace | waypoint | destination ztunnel |
| yes | `targetRefs` | Service | client pod | waypoint |
| yes | `targetRefs` | Gateway | client pod | waypoint |

† Whether or not there is a waypoint in the traffic path.
```

## HTTP Enforcement

In a scenario where policy requires application layer aware attributes, such as HTTP verbs, a waypoint proxy is required.

This table is based on the following invariants:

1. The source pod is a normal pod which has ztunnel enabled.
1. Redirection to the waypoint is configured correctly.
1. The waypoint is configured with the `istio.io/waypoint-for` label set to `service`.

> | Waypoint* | Attachment Style | Resources | Source Identity | Enforced By |
> | --- | --- | --- | --- | --- |
> | no | Selector | Pod | n/a | DENY destination ztunnel |
> | yes | Selector | Pod | n/a | DENY destination ztunnel |
> | no | | Namespace | n/a | DENY destination ztunnel |
> | yes | | Namespace | n/a | DENY destination ztunnel |
> | yes | `targetRefs` | Service | client pod | waypoint |
> | yes | `targetRefs` | Gateway | client pod | waypoint |
> 
> * Whether or not there is already a waypoint is in the traffic path.

Reviewer 2: ```suggestion
| Waypoint † | Attachment Style | Resources | Source Identity | Enforced By |
| --- | --- | --- | --- | --- |
| no | Selector | Pod | n/a | destination ztunnel (always DENY) |
| yes | Selector | Pod | n/a | destination ztunnel (always DENY) |
| no | | Namespace | n/a | destination ztunnel (always DENY) |
| yes | | Namespace | n/a | destination ztunnel (always DENY) |
| yes | `targetRefs` | Service | client pod | waypoint |
| yes | `targetRefs` | Gateway | client pod | waypoint |

† Whether or not there is a waypoint in the traffic path.
```

Why are rows 2 and 4 not "waypoint" in the final column?

---
title: Policy enforcement in ambient mode
description: The two enforcement points for policy in an ambient mesh.
weight: 20
owner: istio/wg-networking-maintainers
test: no
---

> Istio's ambient data plane splits the data plane into two different sets of components. This architecture allows users to only pay for application layer processing if they need it. The trade off made is the extra complexity of understanding how traffic flows through this architecture and where policy will be enforced. This guide will introduce a number of broad scenarios and explore the permutations therein and the details to understand for each scenario.

Reviewer 3: ```suggestion
Istio's ambient data plane splits the data plane into two different sets of components. This architecture allows users to only pay for application layer processing if they need it. The trade-off made is the extra complexity of understanding how traffic flows through this architecture and where policy will be enforced. This guide will introduce a number of broad scenarios and explore the permutations therein and the details to understand for each scenario.
```

## TCP Enforcement

> In the simplest enforcement scenario you simply want to enforce policy against TCP attributes and you have no waypoint proxies in your traffic's path. These policies can be enforced by the ztunnel proxies.

Reviewer 3: ```suggestion
In the simplest enforcement scenario, you simply want to enforce policy against TCP attributes and you have no waypoint proxies in your traffic's path. These policies can be enforced by the ztunnel proxies.
```

> Once you introduce a waypoint proxy the ideal place to enforce policy shifts. Traffic arriving at the destination ztunnel will be coming from the waypoint's identity because waypoint proxies do not impersonate src identity on behalf of the client. This means that even if you only wish to enforce policy against TCP attributes you should bind that policy to your waypoint proxy. A futher TCP policy may be applied to your workload to request that ztunnel enforce things like, "in-mesh traffic needs to come from my waypoint in order to reach my application". This type of policy allows you to choose if "bypassing" the waypoint proxy is permissable in your scenario.

Reviewer 3: ```suggestion
Once you introduce a waypoint proxy, the ideal place to enforce policy shifts. Traffic arriving at the destination ztunnel will be coming from the waypoint's identity because waypoint proxies do not impersonate `src` identity on behalf of the client. This means that even if you only wish to enforce policy against TCP attributes, you should bind that policy to your waypoint proxy. A further TCP policy may be applied to your workload to request that ztunnel enforce things like, "in-mesh traffic needs to come from my waypoint in order to reach my application". This type of policy allows you to choose if "bypassing" the waypoint proxy is permissible in your scenario.
```

| no | | Namespace | client pod | destination ztunnel |
| yes | | Namespace | waypoint | destination ztunnel |
| yes | TargetRef | Service | client pod | waypoint |
| yes | TargetRef | Gateway | client pod | waypoint |

* Whether or not there is already a waypoint is in the traffic path.

## HTTP Enforcement

In a scenario where policy requires application layer aware attributes, such as HTTP verbs, a waypoint proxy is required.

This table is based on the following invariants:

1. The source pod is a normal pod which has ztunnel enabled.
1. Redirection to the waypoint is configured correctly.
1. The waypoint is configured with the `istio.io/waypoint-for` label set to `service`.

| Waypoint* | Attachment Style | Resources | Source Identity | Enforced By |
| --- | --- | --- | --- | --- |
| no | Selector | Pod | n/a | DENY destination ztunnel |
| yes | Selector | Pod | n/a | DENY destination ztunnel |
| no | | Namespace | n/a | DENY destination ztunnel |
| yes | | Namespace | n/a | DENY destination ztunnel |
| yes | TargetRef | Service | client pod | waypoint |
| yes | TargetRef | Gateway | client pod | waypoint |

* Whether or not there is already a waypoint is in the traffic path.

> 
> 

Reviewer 3: ```suggestion

```

## Ingress and Policy Enforcement

> If your application is exposed outside the cluster via and Istio ingress gateway there are additional considerations to be aware of expecially concerning enforcement of TCP layer policy in ztunnel.

Reviewer 3: ```suggestion
If your application is exposed outside the cluster via an Istio ingress gateway, there are additional considerations to be aware of especially concerning enforcement of TCP layer policy in ztunnel.
```
